### PR TITLE
Add clahe image op

### LINKED ISSUE #2359: Contrast Limited AHE (CLAHE)
TFA already includes a global histogram equalization function (`tfa.image.equalize`). I'm proposing adding a python implementation of CLAHE along side that.

**Relevant information**
- Are you willing to contribute it (yes/no): Yes
- Are you willing to maintain it going forward? (yes/no): Yes
- Is there a relevant academic paper? (if so, where): Yes
    - [S. M. Pizer, E. P. Amburn, J. D. Austin, et al.: Adaptive Histogram Equalization and Its Variations. Computer Vision, Graphics, and Image Processing 39 (1987) 355-368.](https://www.cs.unc.edu/Research/Image/MIDAG/pubs/papers/Adaptive%20Histogram%20Equalization%20and%20Its%20Variations.pdf)
- Is there already an implementation in another framework? (if so, where): Yes
    - [OpenCV](https://docs.opencv.org/master/d5/daf/tutorial_py_histogram_equalization.html)
    - [ImageMagick](https://imagemagick.org/script/clahe.php)
- Was it part of tf.contrib? (if so, where): No

**Which API type would this fall under (layer, metric, optimizer, etc.)**
image

**Who will benefit with this feature?**
This specific feature was requested a few years ago in tensorflow main (https://github.com/tensorflow/tensorflow/issues/24181), but it looks like no one has taken it on yet. It's especially useful for enhancing features on some medical imaging data.

**Any other info.**
Originally asked for guidance on the tensorflow developers mailing list and was directed to open a PR here. I have a draft already and it looks like it will fit into tensorflow addons pretty well.

### PR DESCRIPTION
**Pull request by isears**
# Description

Brief Description of the PR:

Fixes #2359 

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [x] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [x] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

If you're adding a bugfix or new feature please describe the tests that you ran to verify your changes:
1. `test_clahe_with_equalize`: tests that the `tfa.image.equalize` function and `tfa.image.clahe` return similar results when the image is untiled and no histogram clipping is performed.
2. `test_clahe_dtype_shape`: tests that function can handle a variety of dtypes and shapes.

### CODE REVIEW

**Comment by WindQAQ** [comment]
Hi @isears, thanks for the contribution. Do you have numpy or c++ equivalent implementation so that I can refer to?

**Comment by bhack** [comment]
https://github.com/scikit-image/scikit-image/blob/master/skimage/exposure/_adapthist.py#L98

**Comment by bhack** [comment]
We have also a TF1 official implementation of the Multidimensional version of CLAHE at https://github.com/VincentStimper/mclahe

**Comment by isears** [comment]
@WindQAQ thanks for looking into this.

In addition to the sklearn implementation mentioned by @bhack, there's an OpenCV C++ implementation [here](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/clahe.cpp).

Here's a side-by-side comparison of all three implementations:
image

Unfortunately, when I check for equality with `np.testing.assert_array_equal()`, all three methods differ substantially (about 90% mismatched elements), even though they arrive at roughly the same results. I think there must be subtle differences in the ways each library handles the underlying mathematics operations. If that's the case, building a robust test against another implementation will be difficult.

Currently, I have a test which uses a single tile with no histogram clipping to imitate normal histogram equalization which seems to agree with the preexisting `tfa.image.equalize()` function. I'm open to suggestions for further test cases.

**Comment by bhack** [comment]
@isears You can also use  `np.testing.assert_allclose` if it is just a small tolerance tuning. If you search have this in many tests.

**Comment by bhack** [comment]
Also can you share the result of a short benchmark of your implementation vs Opencv and Scipy?

**Comment by isears** [comment]
> @isears You can also use `np.testing.assert_allclose` if it is just a small tolerance tuning. If you search have this in many tests.

I tried that as well. I think there's too much variability to use `np.testing.assert_allclose` with any useful margins in a way that wouldn't flake.

> Also can you share the result of a short benchmark of your implementation vs Opencv and Scipy?

Check out  this [Google Colab Notebook](https://colab.research.google.com/drive/1_2IXECe0_ia89uU5rhiyqMa5yLykUm0p?usp=sharing). Looks like ~500 ms for tensorflow, ~5 ms for OpenCV, and ~50 ms for scikit-image. (Also included examples of `np.testing.assert_allclose`)

**Comment by bhack** [comment]
10x over scikit-image it seems to me a quite relevant overhead.

**Comment by isears** [comment]
> 10x over scikit-image it seems to me a quite relevant overhead.

Yeah, I agree. I'm going to take a few days to compare this to the scikit-image implementation and see if I can get better performance out of it.

Maybe making things more similar will also make it possible to implement tests against scikit-image with `numpy.testing.assert_allclose`

**Comment by bhack** [comment]
Yes other than this I suggest also to inspect the compiled program to see if some ops in your composition, that could be fuses, was not fused. See https://www.tensorflow.org/xla#inspect_compiled_programs

**Comment by isears** [comment]
@bhack It looks like the major issue was the histogram building step. In the initial commit, I used `tf.reduce_sum(tf.one_hot())` to build histograms, but in the most recent commit I've replaced that with `tf.math.bincount()`

With those changes, the tensorflow function runs on average 100 ms vs. skimage's 50 ms in the [google colab benchmark](https://colab.research.google.com/drive/1_2IXECe0_ia89uU5rhiyqMa5yLykUm0p?usp=sharing) (a single-cpu environment).

However, `tf.math.bincount()` does not appear to be XLA compatible (i.e. I get errors when I try to compile code containing `tf.math.bincount()` with XLA optimizations). As such, its GPU performance is not as good as `tf.reduce_sum(tf.one_hot())`. On an RTX 2070 Super, for example, I get 20 ms per loop while using `tf.math.bincount()` whereas XLA-compiled `tf.reduce_sum(tf.one_hot())` gets 0.5 ms. 

Which direction should we take this?
1. Optimize for XLA compatibility and GPU performance with `tf.reduce_sum(tf.one_hot())`?
2. Optimize for CPU performance with `tf.math.bincount()`? (and continue to mirror the skimage implementation)
3. Or maybe there's a way to get the best of both worlds?

**Comment by bhack** [comment]
Have you tried to check the IR?  Is `tf.reduce_sum(tf.one_hot())` not fused on CPU?

**Comment by bhack** [comment]
@isears Also if in TF1, but I think it is quite portable to TF2, and multidimensional have you tried to benchmark MCLAHE approach at https://github.com/VincentStimper/mclahe/tree/master/mclahe? (/cc @VincentStimper  @Rea
lPolitiX)?
Reference Pseudocode at pag. 4 https://pure.mpg.de/rest/items/item_3082776/component/file_3181957/content

File: tensorflow_addons/image/color_ops.py [hunk 1]
```diff
@@ -132,3 +133,172 @@ def sharpness(
...
+
+def _clahe(
+    image: TensorLike, clip_limit: Number, tile_grid_size: Union[List[int], Tuple[int]]
+) -> tf.Tensor:
+    """Implements CLAHE as tf ops"""
+    original_2d_shape = (tf.shape(image)[0], tf.shape(image)[1])
+    original_dtype = image.dtype
+
+    # Need image in int32 format for later gather_nd ops
+    image = tf.cast(image, tf.int32)
```
Comment (bhack) [inline tensorflow_addons/image/color_ops.py:146]: Are you sure that having a so early casting to `tf.int32` is ok?

Comment (isears) [reply tensorflow_addons/image/color_ops.py:146]: The first thing that requires a cast to int32 is `tf.math.bincount` (similar to [tfa.image.equalize](https://github.com/tensorflow/addons/blob/77aa837bba0fb88d5a5cb1018c9fcc162726370e/tensorflow_addons/image/color_ops.py#L34)). It could be pushed later, but doing so doesn't seem to affect performance.

If you would like the cast pushed later for other reasons, though, I can definitely push that change.

**Comment by isears** [comment]
> Have you tried to check the IR? Is `tf.reduce_sum(tf.one_hot())` not fused on CPU?

Yes, upon inspection of the compiled XLA I found that you are correct: `one_hot` and `reduce_sum` are fused on the GPU, but not the CPU. Is there anything I can do to force fusion on the CPU?

> @isears Also if in TF1, but I think it is quite portable to TF2, and multidimensional have you tried to benchmark MCLAHE approach

I've benchmarked mclahe [here](https://colab.research.google.com/drive/1AxPC5y5HB81T1FHvMq8olZMDjShT6laR?usp=sharing). On the CPU, it's running at about 1000 ms/loop vs. 100 ms/loop for the current clahe implementation in this PR. The ratio is similar on the GPU. Looking at the pseudo code, it seems like mclahe requires quite a bit more padding than regular clahe, which is probably contributing to the 10x slowdown.

I would be open to taking a stab at porting mclahe into a separate `tfa.image.mclahe` function after this PR gets merged in.

**Comment by bhack** [comment]
@isears  I suggest to post about this at https://groups.google.com/g/xla-dev

**Comment by isears** [comment]
I've posted a topic on the mailing list here: https://groups.google.com/g/xla-dev/c/RrW0bEdi930

**Comment by bhack** [comment]
I suggest you to open a ticket in then main Tensorflow repository of you don't get a reply from the xla mailing list.
It will be assigned to the xla team with an XLA label

**Comment by innat** [comment]
Any update?

**Comment by bhack** [comment]
As we  already have 3 `equalize` implementations in the ecosystem we need to think about a consolidation on these images (and other duplicated) ops /cc @yarri-oss:
https://github.com/tensorflow/models/blob/master/official/vision/beta/ops/augment.py#L580
https://github.com/tensorflow/models/blob/master/official/vision/image_classification/augment.py#L480

**Comment by isears** [comment]
> Any update?

I haven't been able to make contact with the XLA team, but I have implemented this as a pip package here: https://github.com/isears/tf_clahe for anyone who would like to go ahead and use it.

**Comment by bhack** [comment]
@yarri-oss How we could contact the XLA/MLIR/Compile team for these cases?

**Comment by bhack** [comment]
As equalization [was "duplicated" in keras-cv[(https://github.com/keras-team/keras-cv/pull/91) as a preprocessing layers can you try to propose this in the keras-cv repo?

Thanks for the contribution.
